### ISSUE TITLE
How does orange encode categorical values for a neural network? What function does it do and how?

### ISSUE THREAD

**Issue by mlitvinov** [biolab/orange3#3948]
How does orange encode categorical values for a neural network? What function does it do and how?

**Comment by mlitvinov** [comment]
I can not find it in the code

**Comment by markotoplak** [comment]
See `Continuize` from `Orange/preprocess/preprocess.py`. This is executed on the Orange table before it gets passed forward. You will see that categorial values are encoded as indicator variables.

The other preprocessors for scikit-learn models are listed in `SklLearner.preprocessors` (`Orange/base.py`).

**Comment by mlitvinov** [comment]
Can I see in Orange how my data has changed after preprocessing for NN?

**Comment by markotoplak** [comment]
I do not think you can easily see exactly how the data on the input to NN looks. What you can do instead is to use the Preprocess widget and chose equivalent preprocessors there.

**Comment by mlitvinov** [comment]
Do I understand correctly that before loading into a neural network: 1) the missing values ​​of the string format are filled with N / A and the missing values ​​of the number format are filled with an average value 2) All categorical values ​​are encoded as dummy variables.  3) all data is normalized using z-score?  Is this the whole data preprocessing, which is performed by default for a neural network?

**Comment by mlitvinov** [comment]
> I do not think you can easily see exactly how the data on the input to NN looks. What you can do instead is to use the Preprocess widget and chose equivalent preprocessors there.

Do I understand correctly that before loading into a neural network: 1) the missing values ​​of the string format are filled with N / A and the missing values ​​of the number format are filled with an average value 2) All categorical values ​​are encoded as dummy variables. 3) all data is normalized using z-score? Is this the whole data preprocessing, which is performed by default for a neural network?

**Comment by mlitvinov** [comment]
I think I need to explain why I'm asking. The fact is that Orange gives good results in the neural network, but when I try to repeat this through a script, the results become much worse. So I’m looking for a way how I can process them correctly

**Comment by markotoplak** [comment]
It goes like this:
1) Instances with unknown classes are dropped (HasClass preprocessor).
2) All ategorical values are encoded as dummy variables, and all variables are normalized to zero mean and stdev=1 (Continuize).
3) Columns with only unknown values are removed (RemoveNaNColumns).
4) Unknown values are imputed with average values (SklImpute).

The easiest way to see how exactly is your data processed is to modify the Learner class `Orange/base.py` so that its preprocess method is this one:

    def preprocess(self, data):
        """Apply the `preprocessors` to the data"""
        for pp in self.active_preprocessors:
            data = pp(data)
        data.save("test.tab")  # this is the added line
        return data

Then, each training of the model will save the file that NN was trained on.

**Comment by mlitvinov** [comment]
Thank you very much, Marko!
